runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -841,6 +847,238 @@ public BundleFinalizer bundleFinalizer() {
     }
   }
 
+  /**
+   * A concrete implementation of {@link DoFnInvoker.ArgumentProvider} used for running a {@link
+   * DoFn} on window expiration.
+   */
+  private class OnWindowExpirationArgumentProvider extends DoFn<InputT, OutputT>.OnTimerContext
```

Comment:
Why is this based on OnTimerContext?

Reply:
@reuvenlax OnWindowExpirationContext added.